### ISSUE TITLE
worflow REST API returns empty file list

### ISSUE THREAD

**Issue by belforte** [dmwm/CRABServer#6595]
There is a problem in new CRABServer
https://hypernews.cern.ch/HyperNews/CMS/get/computing-tools/6046/1/1.html
I have rolled back production to previous release v3.210110 via
./scripts/deploy-srv.sh crabserver HG2102d

Problem can be reproduced also in test instance, I will use it to debug.
A quick way to check is to lookup this URL in the browser

https://cmsweb.cern.ch:8443/crabserver/prod/workflow?workflow=210513_214945%3Ajschulte_crab_dileptonAna_muons_2016_Wantitop&subresource=data2&limit=-1&jobids=2

which returns [1]

and comparing with the same on cmsweb-testbed which runs the new versions

the problem with the new version is that output is simply an empty dictionary [2]
no error is logged anywhere.
I will use test2 to debug this.

[1]
{"result": [
 {"size": 2175707, "tmpsite": "T2_IT_Legnaro", "tmplfn": "/store/user/jschulte/ST_tW_antitop_5f_inclusiveDecays_13TeV-powheg-pythia8_TuneCUETP8M1/dileptonAna_muons_2016_Wantitop/210513_214945/0000/zp2mu_histos_2.root", "directstageout": "F", "checksum": {"adler32": "104efc2f", "cksum": 2602359556, "md5": "104efc2f"}, "lfn": "/store/user/jschulte/ST_tW_antitop_5f_inclusiveDecays_13TeV-powheg-pythia8_TuneCUETP8M1/dileptonAna_muons_2016_Wantitop/210513_214945/0000/zp2mu_histos_2.root", "site": "T2_US_Purdue", "jobid": "2"}
]}

[2]
{"result": [
]}

**Comment by belforte** [comment]
test URL on dev DB:
https://cmsweb-test1.cern.ch/crabserver/dev/workflow?workflow=210513_204322:belforte_crab_20210513_224314&subresource=data2&limit=1&jobids=1
works because I have put5 v3.210110 on cmsweb-test1
```
belforte@lxplus806/TC3> X5curl "https://cmsweb-test1.cern.ch/crabserver/dev/info"
{"desc": {"columns": ["null"]}, "result": [
 {"crabserver": "Welcome", "version": "v3.210110"}
]}
belforte@lxplus806/TC3> X5curl "https://cmsweb-test1.cern.ch/crabserver/dev/workflow?workflow=210513_204322:belforte_crab_20210513_224314&subresource=data2&limit=1&jobids=1"
{"result": [
 {"size": 41583514, "tmpsite": "T1_IT_CNAF_Disk", "tmplfn": "/store/user/belforte/GenericTTbar/Stefano-Test-210506/210513_204322/0000/kk_1.root", "directstageout": "F", "checksum": {"adler32": "2f5539e9", "cksum": 558031825, "md5": "2f5539e9"}, "lfn": "/store/user/belforte/GenericTTbar/Stefano-Test-210506/210513_204322/0000/kk_1.root", "site": "T2_CH_CERN", "jobid": "1"}
]}
```

while on cmsweb-test2:
```
belforte@lxplus806/TC3> X5curl "https://cmsweb-test2.cern.ch/crabserver/dev/info"
{"desc": {"columns": ["null"]}, "result": [
 {"crabserver": "Welcome", "version": "development"}
]}
belforte@lxplus806/TC3> X5curl "https://cmsweb-test2.cern.ch/crabserver/dev/workflow?workflow=210513_204322:belforte_crab_20210513_224314&subresource=data2&limit=1&jobids=1"
{"result": [
]}
```

**Comment by belforte** [comment]
the problem comes from this lines
https://github.com/dmwm/CRABServer/blob/5934e165b8072256a09f8afd2c6ecadebe34e589/src/python/CRABInterface/HTCondorDataWorkflow.py#L113-L116
because `jobids` are list of bytes already
```
(Pdb) jobids
[b'1']
```
due to
https://github.com/dmwm/CRABServer/blob/01c8b0c37a7f9230caae963adab5b78716dac68f/src/python/CRABInterface/RESTUserWorkflow.py#L488
and when executing `jobids = [str(intJobId) for intJobId in jobids]` one obtains
```
(Pdb) jobids
["b'1'"]
```

**Comment by belforte** [comment]
replacing `validate_strlist` with `validate_numlist` in https://github.com/dmwm/CRABServer/blob/01c8b0c37a7f9230caae963adab5b78716dac68f/src/python/CRABInterface/RESTUserWorkflow.py#L531 indeed solves this, but I am worried that things may break for automatic splitting when jobids need to be strings.
I suspect that https://github.com/dmwm/CRABServer/blob/5934e165b8072256a09f8afd2c6ecadebe34e589/src/python/CRABInterface/HTCondorDataWorkflow.py#L116 is completely un-necessary and worked by chance. Even in old CRABServer version, `jobids` at this point is already a list of strings !
Will double check the old versions anyhow

**Comment by belforte** [comment]
indeed in CRABServer v3.210110, which was built with the old pre-python3 WMCore, the line
https://github.com/dmwm/CRABServer/blob/5934e165b8072256a09f8afd2c6ecadebe34e589/src/python/CRABInterface/HTCondorDataWorkflow.py#L116
has absolutely no effect, in my example `jobids` is set to `['1']` both before and after that.

**Comment by belforte** [comment]
So I will remove it, and we will need to fully validate again (and better), because there can always be edge cases. At the very least I will try an automatic splitting task.

**Comment by belforte** [comment]
By the way `crab getoutput` is part of current CRAB Client validation test
https://github.com/ddaina/Docker/blob/4dc0260720c32356aa6c731ccbe5dcee09b60506/crab_tests/CRABClientTests/CRABClientScripts/CRABClientTests.sh#L208
@ddaina can you find out why we did not spot this problem ? This independent of CRAB Client version and AFAIU it should have happened anytime we ran the test with an CRABServer built after "February". Did we simply not run it ? Or is the test flawed ?

**Comment by belforte** [comment]
fix tagged in v3.210514

**Comment by belforte** [comment]
> By the way `crab getoutput` is part of current CRAB Client validation test
> https://github.com/ddaina/Docker/blob/4dc0260720c32356aa6c731ccbe5dcee09b60506/crab_tests/CRABClientTests/CRABClientScripts/CRABClientTests.sh#L208
> @ddaina can you find out why we did not spot this problem ? This independent of CRAB Client version and AFAIU it should have happened anytime we ran the test with an CRABServer built after "February". Did we simply not run it ? Or is the test flawed ?

I think I know the answer. The CRABClientTests.sh only checks that command ended with exit code 0, not that it did what was expected. In this case command terminated w/o error, but retrieved no files, like it does when e.g. all jobs are still running.
No obvious general solution, we need to be aware, while thinking on how to do better
